src/cargo/ops/cargo_package.rs

```diff
@@ -185,44 +185,52 @@ fn infer_registry(
     pkgs: &[&Package],
     reg_or_index: Option<RegistryOrIndex>,
 ) -> CargoResult<SourceId> {
-    let publish_registry = if let Some(RegistryOrIndex::Registry(registry)) = reg_or_index.as_ref()
-    {
-        Some(registry.clone())
-    } else if let Some([first_pkg_reg]) = pkgs[0].publish().as_deref() {
-        // If no registry is specified in the command, but all of the packages
-        // to publish have the same, unique allowed registry, push to that one.
-        if pkgs[1..].iter().all(|p| p.publish() == pkgs[0].publish()) {
-            Some(first_pkg_reg.clone())
-        } else {
-            None
+    let reg_or_index = match reg_or_index {
+        Some(r) => r,
+        None => {
+            if pkgs[1..].iter().all(|p| p.publish() == pkgs[0].publish()) {
+                // If all packages have the same publish settings, we take that as the default.
+                match pkgs[0].publish().as_deref() {
+                    Some([unique_pkg_reg]) => RegistryOrIndex::Registry(unique_pkg_reg.to_owned()),
+                    None | Some([]) => RegistryOrIndex::Registry(CRATES_IO_REGISTRY.to_owned()),
+                    _ => bail!(
+                "cannot infer a registry\n\
+                Multiple `package.publish` settings found. Use --registry or --index to choose one."
+            ),
+                }
+            } else {
+                bail!(
+            "cannot infer a registry\n\
+            Not all packages have the same `package.publish` setting. Use --registry or --index to choose one."
```

Comment:
I feel like these messages aren't entirely clear from the end-users perspective, especially for running `cargo package`.  I'm just not sure how to entirely solve it if `cargo publish` will end up going through this code path.

At a high level, for the first message, I'd focus on it being ambiguous and I'd name the choices.  Maybe "`--registry is required to disambiguate between A, B, or C registries". Yes, this doesn't mention `--index` maybe that could be an "or" in there.

For the second, the issue is that it is either ambiguous (there is a common registry but other registries as well) or there is a conflict (for which `--registry` won't help).

Reply:
>  I'm just not sure how to entirely solve it if cargo publish will end up going through this code path.

If we end up needing different error messages, my plan was to factor out a function that returns an enum of the possible inference errors. But it will indeed be easier if the errors can make sense for both `package` and `publish`.

I added a distinction between "non-overlapping `publish` fields" and "not-all-equal `publish` fields", but I haven't managed to come up with a great error message. I remember there was some discussion about diagnostics that point to locations in manifest files. Does that exist yet? Because I think that if we could point to the multiple distinct `publish` fields then it will be clear even if the message isn't great.